Workflow/Tests/WorkflowObserverTests.swift

```diff
@@ -0,0 +1,537 @@
+            StateTransitioningWorkflow.Event.toggle,
+            workflow: StateTransitioningWorkflow(),
+            session: .testingSession
+        )
+
+        XCTAssertEqual(
+            callbackSequence,
+            ["one", "two"]
+        )
+    }
+
+    func test_chainedObserver_willApplyActionCallbacks() {
+        var willApplyCallbackSequence: [String] = []
+        var didApplyCallbackSequence: [String] = []
+
+        let observer1 = TestObserver()
+        observer1.onApplyAction = { _, _, _, _ in
+            willApplyCallbackSequence.append("one")
+            return { _, _ in
+                didApplyCallbackSequence.append("one")
+            }
+        }
+
+        let observer2 = TestObserver()
+        observer2.onApplyAction = { _, _, _, _ in
+            willApplyCallbackSequence.append("two")
+            return { _, _ in
+                didApplyCallbackSequence.append("two")
+            }
+        }
+
+        let chained = ChainedWorkflowObserver(observers: [observer1, observer2])
+        let didApply = chained.workflowWillApplyAction(
+            StateTransitioningWorkflow.Event.toggle,
+            workflow: StateTransitioningWorkflow(),
+            state: false,
+            session: .testingSession
+        )
+
+        XCTAssertEqual(willApplyCallbackSequence, ["one", "two"])
+        XCTAssertEqual(didApplyCallbackSequence, [])
+
+        didApply?(true, nil)
+
+        XCTAssertEqual(didApplyCallbackSequence, ["two", "one"])
+    }
+}
+
+// MARK: - Integration Points
+
+extension WorkflowObserverTests {
+    func test_workflowHost_chainsObservers() {
+        // 0 observers
+        do {
+            let host = WorkflowHost(
+                workflow: Parent(),
+                observers: nil
+            )
+
+            XCTAssertNil(host.rootNode.observer)
+        }
+
+        // 1 observer
+        do {
+            let host = WorkflowHost(
+                workflow: Parent(),
+                observers: [TestObserver()]
+            )
+
+            XCTAssert(host.rootNode.observer is TestObserver)
+        }
+
+        // > 1 observer
+        do {
+            struct NoOpObserver: WorkflowObserver {}
+
+            let host = WorkflowHost(
+                workflow: Parent(),
+                observers: [TestObserver(), NoOpObserver()]
+            )
+
+            let observer = host.rootNode.observer as? ChainedWorkflowObserver
+            XCTAssertNotNil(observer)
+
+            XCTAssertEqual(observer?.observers.count, 2)
+            XCTAssert(observer?.observers.first is TestObserver)
+            XCTAssert(observer?.observers.last is NoOpObserver)
+        }
+    }
+}
+
+// MARK: - Utilities
+
+private final class TestObserver: WorkflowObserver {
+    var onSessionBegan: ((WorkflowSession) -> Void)?
+    var onSessionEnded: ((WorkflowSession) -> Void)?
+    /// (Workflow, State, Session) -> Void
+    var onDidMakeInitialState: ((Any, Any, WorkflowSession) -> Void)?
```

Comment:
What do you think about making this class generic over `WorkflowType`?

Reply:
i'm not sure if that would help anything, since the callbacks can run on heterogenous Workflow types. i agree this doesn't seem like the most elegant approach, but it seems fairly useful as a testing device.

Reply:
>the callbacks can run on heterogenous Workflow types

Yeah that's what I missed all around 👍